### ISSUE TITLE
search in relation fields does not work properly

### ISSUE THREAD

**Issue by Seyolas** [strapi/strapi#18257]
### Required System information
- Node.js version: 18.14.2
- NPM version: 9.5.0
- Strapi version:4.12.5
- Database:postgres
- Operating system: windows
- Is your project Javascript or Typescript: js

### Describe the bug

![Screenshot 2023-10-03 151056](https://github.com/strapi/strapi/assets/39236950/3c0cb9a7-974c-48fb-9691-dfca8245dda1)
![Screenshot 2023-10-03 151106](https://github.com/strapi/strapi/assets/39236950/2777297b-fcba-4206-802a-15b5d9b6dbd5)
![Screenshot 2023-10-03 151115](https://github.com/strapi/strapi/assets/39236950/35804bf7-527c-4e91-821c-bcfaf15d483b)
![Screenshot 2023-10-03 151127](https://github.com/strapi/strapi/assets/39236950/dbfb8fa5-60de-4700-b6a1-4bb28dc4d047)

### Expected behavior
it would be nice to see data when each letter is typed (or maybe at least 3 letter)

**Comment by alex-brukhty** [comment]
I have a similar issue on v4.14.2, when typing a word that is not first in potential results, then **no results** in the field dropdown.
But I can see all the results in the API response for the field if inspecting the network tab.

**Comment by akshayw1** [comment]
Could you pls assign it to me

**Comment by hosnas** [comment]
is there any ETA for this?

**Comment by derrickmehaffy** [comment]
Raising sev due to to internal TID4719

**Comment by derrickmehaffy** [comment]
The issue is most likely we are using `$startsWithi` instead of `$contains`

![image](https://github.com/strapi/strapi/assets/8593673/5f529ae8-c269-40fc-b004-bfbd1497c0f4)

**Comment by derrickmehaffy** [comment]
Source of the issue is here: https://github.com/strapi/strapi/blob/8ec354cd1dac7ebd1e675a82238d1538ac678f52/packages/core/admin/admin/src/content-manager/hooks/useRelation/useRelation.js#L169

**Comment by derrickmehaffy** [comment]
Problem originally caused by: https://github.com/strapi/strapi/pull/18110

**Comment by t-fritsch** [comment]
I don't know if a fix is on his way but it would be great if this setting wasn't hard coded or if we had a way to control it easily by code 🙏

**Comment by derrickmehaffy** [comment]
> I don't know if a fix is on his way but it would be great if this setting wasn't hard coded or if we had a way to control it easily by code 🙏

I've raised it internally with the team but waiting on engineering to make a decision on it at the moment.

**Comment by t-fritsch** [comment]
OK thanks a lot for the feedback ! (_If I can add a little input on this, this bug is particularly annoying for large databases : for example our relation target table has ~130k records so startsWith is just too restrictive to be useful_)

**Comment by derrickmehaffy** [comment]
Funnily enough that's why it was changed in the PR but that PR should not have been approved as it was a breaking change. The startsWith is "easier" on the DB for performance but contains is more useful.

**Comment by t-fritsch** [comment]
following a discussion on the strapi discord server, I'd like to add that if the filtering was only done server side, it would allow anyone to customise this behavior with a global middleware by simply replacing `_filter=$startsWithi` by `_filter=$containsi` (or anything else) in the `ctx.request.search` string. Not ideal, but at least a workaround.

Following this blog post https://strapi.io/blog/understanding-the-strapi-request-flow-a-journey-from-koa-to-modern-middleware-architecture, I made a quick test and it "half" works : 
```ts
ctx.request.search = ctx.request.search.replace('_filter=$startsWithi', '_filter=$containsi');
await next();
```
The AJAX call made by the relation combobox to `/content-manager/relations/api::mymodel.mymodel/myrelation?entityId=...&pageSize=10&_q=mysearchstring&_filter=$startsWithi&page=1` returns every record that "contains" `mysearchstring`, not only those "starting with" the string 🥳

The problem is that with current version of strapi the data returned by the API is not shown in the relation combobox, since the front also adds his own filtering pass with startsWith 🙁

I wonder if this double data filtering is really needed.

**Comment by derrickmehaffy** [comment]
No it's not needed for sure, that's where the discussion was at on why we implemented it in the frontend.

**Comment by derrickmehaffy** [comment]
Discussion is still ongoing but it's out of my hands, at the moment either @Aurelsicoko , @alexandrebodin , or @joshuaellis need to take the decision on when this going to happen.

I do believe we have agreed that the intended functionality should be using the insensitive case contains filter should be used instead of the starts with but as for timeline that's beyond my job/team's ability to say. I'm on our support team who do the initial triage of issues but as for fix implementation and scheduling that's up to our product and engineering teams with Aurelien being our CPO and Alex being our CTO and Josh being the engineer who knows this part of the code the best.

I would like us to fix this soon as it's impacting several users including some of our enterprise and cloud customers as well.

There was a gentle nudge shared earlier today by our sales team so I'm hoping for an official deadline soon ish.

Sorry for the delay, I am trying my best here though.

**Comment by t-fritsch** [comment]
Since it couldn't be done in 4.15.5, is there any chance that this gets added to 4.15.6 backlog (would be great because the regression was first reported in July #17391) ? 🙏 
This is really hurting customer experience and the fix for this regression should be quick (basically revert `startsWith` and revert it to `contains` value).

**Comment by t-fritsch** [comment]
thanks a lot @joshuaellis ! 🙏

**Comment by joshuaellis** [comment]
Sorry if i wasn't clear enough, i meant it won't be released until next year – afaik we're not doing a release this side of new year :)

**Comment by derrickmehaffy** [comment]
Adding to Josh here I believe the plan is to include this in 4.16.0 scheduled for Jan 3rd 2024 based on the last internal sync we had yesterday on the release schedule.
